File: src/main/java/com/mmust/evaluators/impl/StackPostfixMathExpressionEvaluator.java
package com.mmust.evaluators.impl;

import com.mmust.token.Token;
import com.mmust.evaluators.PostfixEvaluator;
import com.mmust.token.Operator;
import com.mmust.token.NumToken;
import com.mmust.token.Number;

import java.util.Deque;
import java.util.LinkedList;

public class StackPostfixMathExpressionEvaluator implements PostfixEvaluator {

	@Override
	public double eval(Iterable<Token> postfixExpr) {
		Deque<Token> stack = new LinkedList<>();

		for (Token t : postfixExpr) {
			switch (t.getType()) {
				case NUMBER:
					stack.push(t);
					break;

				case OPERATOR:
					Number op2 = (Number) stack.pop();
					Number op1 = (Number) stack.pop();

					Operator evaluator = (Operator) t;

					double result = evaluator.eval(op1.val(), op2.val());

					stack.push(new NumToken(result));
					break;
			}
		}

		Number result = (Number) stack.pop();

		return result.val();
	}
}

File: src/main/kotlin/com/mmust/evaluators/impl/StackPostfixMathExpressionEvaluator.kt
package com.mmust.evaluators.impl

import com.mmust.evaluators.PostfixEvaluator
import com.mmust.token.NumToken
import com.mmust.token.Number
import com.mmust.token.Operator
import com.mmust.token.Token
import java.util.*

class StackPostfixMathExpressionEvaluator : PostfixEvaluator {

    override fun eval(postfixExpr: Iterable<Token>): Double {
        val stack = LinkedList<Token>()

        for (t in postfixExpr) {
            when (t) {
                is Number -> stack.push(t)

                is Operator -> {
                    val op2 = stack.pop() as Number
                    val op1 = stack.pop() as Number

                    val result = t.eval(op1.`val`(), op2.`val`())

                    stack.push(NumToken(result))
                }
            }
        }

        val result = stack.pop() as Number

        return result.`val`()
    }
}
